flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -621,6 +692,237 @@ private static void checkInterrupted() throws InterruptedException {
+				stateBackend.cancelStreamRegistry.registerClosable(inputStream);
+
+				outputStream = checkpointStreamFactory
+					.createCheckpointStateOutputStream(checkpointId, checkpointTimestamp);
+				stateBackend.cancelStreamRegistry.registerClosable(outputStream);
+
+				while (true) {
+					int numBytes = inputStream.read(buffer);
+
+					if (numBytes == -1) {
+						break;
+					}
+
+					outputStream.write(buffer, 0, numBytes);
+				}
+
+				return outputStream.closeAndGetHandle();
+			} finally {
+				if (inputStream != null) {
+					stateBackend.cancelStreamRegistry.unregisterClosable(inputStream);
+					inputStream.close();
+					inputStream = null;
+				}
+
+				if (outputStream != null) {
+					stateBackend.cancelStreamRegistry.unregisterClosable(outputStream);
+					outputStream.close();
+					outputStream = null;
+				}
+			}
+		}
+
+		private StreamStateHandle materializeMetaData() throws Exception {
+			try {
+				outputStream = checkpointStreamFactory
+					.createCheckpointStateOutputStream(checkpointId, checkpointTimestamp);
+				stateBackend.cancelStreamRegistry.registerClosable(outputStream);
+
+				KeyedBackendSerializationProxy serializationProxy =
+					new KeyedBackendSerializationProxy(stateBackend.keySerializer, stateMetaInfos);
+				DataOutputView out = new DataOutputViewStreamWrapper(outputStream);
+
+				serializationProxy.write(out);
+
+				return outputStream.closeAndGetHandle();
+			} finally {
+				if (outputStream != null) {
+					stateBackend.cancelStreamRegistry.unregisterClosable(outputStream);
+					outputStream.close();
+					outputStream = null;
+				}
+			}
+		}
+
+		void takeSnapshot() throws Exception {
+			// use the last completed checkpoint as the comparison base.
+			baseSstFiles = stateBackend.materializedSstFiles.get(stateBackend.lastCompletedCheckpointId);
+
+			// save meta data
+			for (Map.Entry<String, Tuple2<ColumnFamilyHandle, RegisteredBackendStateMetaInfo<?, ?>>> stateMetaInfoEntry : stateBackend.kvStateInformation.entrySet()) {
+
+				RegisteredBackendStateMetaInfo<?, ?> metaInfo = stateMetaInfoEntry.getValue().f1;
+
+				KeyedBackendSerializationProxy.StateMetaInfo<?, ?> metaInfoProxy =
+					new KeyedBackendSerializationProxy.StateMetaInfo<>(
+						metaInfo.getStateType(),
+						metaInfo.getName(),
+						metaInfo.getNamespaceSerializer(),
+						metaInfo.getStateSerializer());
+
+				stateMetaInfos.add(metaInfoProxy);
+			}
+
+			// save state data
+			backupPath = new Path(stateBackend.instanceBasePath.getAbsolutePath(), "chk-" + checkpointId);
+			backupFileSystem = backupPath.getFileSystem();
+			if (backupFileSystem.exists(backupPath)) {
+				LOG.warn("Deleting an existing local checkpoint directory " +
+					backupPath + ".");
+
+				backupFileSystem.delete(backupPath, true);
+			}
+
+			// create hard links of living files in the checkpoint path
+			Checkpoint checkpoint = Checkpoint.create(stateBackend.db);
+			checkpoint.createCheckpoint(backupPath.getPath());
```

Comment:
@StefanRRichter Thanks for your explanation. Your method to allow fast checkpoint is very smart. I think it can help us to avoid small sst files produced by the flushing.  But i think we should also be careful about the flushing of the buffer, because it might block the processing if the buffer is very big.